app/src/main/java/com/example/drawingApp/GradientSquare.kt

```diff
@@ -0,0 +1,176 @@
+package com.example.drawingApp
+
+import android.annotation.SuppressLint
+import android.content.Context
+import android.graphics.*
+import android.util.AttributeSet
+import android.view.MotionEvent
+import android.view.View
+import android.view.ViewGroup
+import androidx.core.view.children
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
+import kotlin.math.max
+import kotlin.math.min
+
+@SuppressLint("ClickableViewAccessibility")
+class GradientSquare @JvmOverloads constructor(
+    context: Context,
+    attrs: AttributeSet? = null,
+    defStyle: Int = 0
+) : View(context, attrs, defStyle) {
+
+    private val lock: Any = Any()
+    private val selectionPaint = Paint().apply {
+        strokeWidth = 5f
+        style = Paint.Style.STROKE
+        color = Color.WHITE
+    }
+
+    private var held = false
+    private lateinit var drawRect: Rect
+
+    /**
+     * called on saturation or value change.
+     * use this to update any saturation of values held elsewhere.
+     */
+    lateinit var onSatOrValChange: (saturation: Float, value: Float) -> Unit
+
+    /**
+     * called when user presses down on the gradient,
+     * used to prevent dragging
+     */
+    var onDown: (() -> Unit)? = null
+
+    /**
+     * called when user lifts off gradient or cancels,
+     * allows the user to drag dialog again
+     */
+    var onUp: (() -> Unit)? = null
+    var sat = 1f
+    var valu = 1f
+    var hue = 0f
+
+    init {
+        setOnTouchListener { _, motionEvent ->
+            sat = max(min(motionEvent.x / width, 1f), 0f)
+            valu = max(min((height - motionEvent.y) / height, 1f), 0f)
+            when (motionEvent.action) {
+                MotionEvent.ACTION_DOWN -> {
+                    held = true
+                    onDown?.invoke()
+                    (this.parent as? ViewGroup)?.children?.forEach { it.clearFocus() }
+                    onSatOrValChange(sat, valu)
+                }
+                MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> {
+                    held = false
+                    onUp?.invoke()
+                }
+                MotionEvent.ACTION_MOVE -> {
+                    if (held) {
+                        onSatOrValChange(sat, valu)
+                    }
+                }
+            }
+            true
+        }
+    }
+
+    private fun startCacheTask() {
+        GlobalScope.launch(Dispatchers.IO) {
+            regenerate()
+            completedIndexes?.let {
+                val startPoint = completedIndexes
+                if (startPoint != null) {
+                    for (i in startPoint..359) {
+                        regenerate()
+                        bitmaps?.set(i, generateBitmap(i))
+                        completedIndexes = completedIndexes?.plus(1)
+                    }
+                    areBitmapsDone = true
```

Comment:
none for the purposes of use, removed one

Reply:
Actually there is a specific line it is important, to tell whether we need to start cacheing, because while in the process of cacheing we are not done, if I only have one bool I'll constantly start cacheing over and over.